### DISCUSSION TITLE
Hide popup on unhover

### DISCUSSION THREAD

**Discussion by knoellle** [emilk/egui#2177] [Q&A]
Hello, I'm trying to make a kind of combobox where the user can toggle different checkboxes on and off. However, when I use an actual combobox, it closes immediately once the user clicks one of the comboboxes.
Using a menu button is better, but I also don't want the user to have to click somewhere else to make the menu disappear. I would like the menu box to disappear when the mouse cursor moves away from the box.
I have tried to manually hide the popup when it loses hover focus, but I'm not sure how to access the correct response to detect the event. Actually closing the menu should just be `ui.close_menu()`, right?

Any ideas or hints on how I could implement something like this without writing my own popup system?

**Comment by willser** [comment]
Could you show some code?I am not sure do I understand this question.

**Reply by knoellle** [reply]
I used to have a combobox:
```rust
ComboBox::from_id_source("Overlays")
    .selected_text("Overlays")
    .show_ui(ui, |ui| {
        ui.checkbox(&mut self.candidates, "Candidates");
        ui.checkbox(&mut self.detections, "Detections");
        ...
    });
```
This has the problem of closing every time the user interacts with a checkbox, which is super annoying when you want to toggle two or more options.

So I tried a menu button instead:
```rust
ui.menu_button("Overlays", |ui| {
    ui.checkbox(&mut self.candidates, "Candidates");
    ui.checkbox(&mut self.detections, "Detections");
    ...
}
```
But this has the opposite problem: It only closes when the user deliberately clicks somewhere else (outside the popup area) or presses the Escape key.

Now, depending on the application, the latter behavior may not be an issue. But in my case, the vast majority of the window space reacts to clicks in some way.

What I want to achieve is behavior from the menu button where toggling a checkbox does *not* close the menu popup, but instead have it close when the mouse cursor leaves the popup area (ideally with some margin around it), so that the user doesn't have to press a button or click somewhere to close it.
Something like this:
```rust
ui.menu_button("Overlays", |ui| {
    ui.checkbox(&mut self.candidates, "Candidates");
    ui.checkbox(&mut self.detections, "Detections");
    ...

    if !cursor_near_popup_area {
        ui.close_menu();
    }
}
```
I'm just not sure how to access the popup menu area to check it for cursor proximity.
The response from the `menu_button` call only seems to describe the button itself, not the popup menu.

**Reply by willser** [reply]
It seems no way without change egui code:

https://github.com/emilk/egui/blob/041f2e64bac778c9095fbf4316dc1f7c7bceb670/crates/egui/src/menu.rs#L283-L286

**Comment by ccmvn** [comment]
Any update about this?

```
fn create_menu_button(ui: &mut Ui, button_name: &str, menu_content: impl FnOnce(&mut Ui)) -> Response {
    let response = ui.button(button_name);

    if response.clicked() {
        if ui.memory(|memory| memory.is_popup_open(response.id)) {
            ui.memory_mut(|memory| memory.close_popup());
        } else {
            ui.memory_mut(|memory| memory.open_popup(response.id));
        }
    }

    if ui.input(|input| input.key_pressed(Key::Escape)) {
        ui.memory_mut(|memory| memory.close_popup());
    }

    if ui.memory(|memory| memory.is_popup_open(response.id)) {
        Area::new(response.id)
            .order(Order::Foreground) // Makes the popup appear over other ui elements
            .fixed_pos(Pos2::new(response.rect.min.x + 5.0, response.rect.max.y + 5.0)) // Add space from the left and bottom
            .show(ui.ctx(), |ui| menu_content(ui));
    }

    return response;
}
```

I need exactly this functionality for my menu
